sklearn/utils/_readonly_array_wrapper.pyx

```diff
@@ -0,0 +1,67 @@
+"""
+ReadonlyArrayWrapper implements the buffer protocol to make the wraped buffer behave as if
+writeable, even for readonly buffers. This way, even readonly arrays can be passed as
+argument of type (non const) memoryview.
+This is a workaround for the missing support for const fused-typed memoryviews in
+Cython < 3.0.
+
+Note: All it does is LIE about the readonly attribute: tell it's false!
+This way, we can use it on arrays that we don't touch.
+!!! USE CAREFULLY !!!
+"""
+# TODO: Remove with Cython >= 3.0 which supports const memoryviews for fused types.
+
+from cpython cimport Py_buffer
+from cpython.buffer cimport PyObject_GetBuffer, PyBuffer_Release, PyBUF_WRITABLE
+
+import numpy as np
+cimport numpy as np
+
+
+np.import_array()
+
+
+ctypedef fused NUM_TYPES:
+    np.npy_float64
+    np.npy_float32
+    np.npy_int64
+    np.npy_int32
+
+
+cdef class ReadonlyArrayWrapper:
+    cdef object wraps
+
+    def __init__(self, wraps):
+        self.wraps = wraps
+
+    def __getbuffer__(self, Py_buffer *buffer, int flags):
+        request_for_writeable = False
+        if flags & PyBUF_WRITABLE:
+            flags ^= PyBUF_WRITABLE
+            request_for_writeable = True
+        PyObject_GetBuffer(self.wraps, buffer, flags)
+        if request_for_writeable:
+            buffer.readonly = False  # This is a lie!
```

Comment:
If `self.wraps` is writable and a writable buffer is requested, this is not actually a lie.

```suggestion
            # This is a lie when self.wraps is readonly
            buffer.readonly = False
```

Reply:
Which makes the lie intriguing:smirk: